microsoft-integration/impl/src/main/java/org/sakaiproject/microsoft/impl/MicrosoftCommonServiceImpl.java

```diff
@@ -1470,7 +1475,100 @@ public List<MeetingRecordingData> getOnlineMeetingRecordings(String onlineMeetin
 
 		return ret;
 	}
-	
+
+	@Override
+	public List<AttendanceRecord> getMeetingAttendanceReport(String onlineMeetingId, String userEmail) throws MicrosoftCredentialsException {
+
+		List<AttendanceRecord> attendanceRecordsResponse = new ArrayList<>();
+
+		MicrosoftUser organizerUser = getUserByEmail(userEmail);
+
+		MeetingAttendanceReportCollectionPage attendanceReports = getGraphClient()
+				.users(organizerUser.getId())
+				.onlineMeetings(onlineMeetingId)
+				.attendanceReports()
+				.buildRequest()
+				.get();
+
+		if (attendanceReports != null && attendanceReports.getCurrentPage() != null) {
+			for (com.microsoft.graph.models.MeetingAttendanceReport report : attendanceReports.getCurrentPage()) {
+				String reportId = report.id;
+
+				AttendanceRecordCollectionPage attendanceRecords = getGraphClient()
+						.users(organizerUser.getId())
+						.onlineMeetings()
+						.byId(onlineMeetingId)
+						.attendanceReports(reportId)
+						.attendanceRecords()
+						.buildRequest()
+						.get();
+
+				if (attendanceRecords != null && attendanceRecords.getCurrentPage() != null) {
+					for (com.microsoft.graph.models.AttendanceRecord record : attendanceRecords.getCurrentPage()) {
+						AttendanceRecord response = new AttendanceRecord();
+
+						if (record.emailAddress != null) response.setEmail(record.emailAddress);
+						if (record.id != null) response.setId(record.id);
+						if (record.identity.displayName != null) response.setDisplayName(record.identity.displayName);
+						if (record.role != null) response.setRole(record.role);
+						if (record.totalAttendanceInSeconds != null) response.setTotalAttendanceInSeconds(record.totalAttendanceInSeconds);
+
+						List<AttendanceInterval> intervals = new ArrayList<>();
+						if (record.attendanceIntervals != null) {
+							for (com.microsoft.graph.models.AttendanceInterval interval : record.attendanceIntervals) {
+								AttendanceInterval intervalResponse = new AttendanceInterval();
+								if (interval.joinDateTime != null) intervalResponse.setJoinDateTime(interval.joinDateTime.toString());
+								if (interval.leaveDateTime != null) intervalResponse.setLeaveDateTime(interval.leaveDateTime.toString());
+								if (interval.durationInSeconds != null) intervalResponse.setDurationInSeconds(interval.durationInSeconds);
+								intervals.add(intervalResponse);
+							}
+						}
+						response.setAttendanceIntervals(intervals);
+
+
+						attendanceRecordsResponse.add(response);
+
+					}
+				}
+
+			}
+
+
+		}
+		return attendanceRecordsResponse;
+	}
+
+
+	@Transactional(readOnly = true)
+	public byte[] createAttendanceReportCsv(List<AttendanceRecord> attendanceRecords, List<String> columnNames) {
+		if (columnNames == null || columnNames.size() != 7) {
```

Comment:
I'll use a constant defined at the top for the number 7. 
public static int COLUMN_SIZE = 7.